titiler/api/endpoints/mosaic.py

```diff
@@ -0,0 +1,401 @@
+@router.get(r"/tiles/{z}/{x}/{y}@{scale}x", **tile_response_codes)
+@router.get(r"/tiles/{z}/{x}/{y}@{scale}x\.{format}", **tile_response_codes)
+async def mosaic_tile(
+    z: int = Path(..., ge=0, le=30, description="Mercator tiles's zoom level"),
+    x: int = Path(..., description="Mercator tiles's column"),
+    y: int = Path(..., description="Mercator tiles's row"),
+    scale: int = Query(
+        1, gt=0, lt=4, description="Tile size scale. 1=256x256, 2=512x512..."
+    ),
+    format: ImageType = Query(None, description="Output image type. Default is auto."),
+    pixel_selection: PixelSelectionMethod = Query(
+        PixelSelectionMethod.first, description="Pixel selection method."
+    ),
+    image_params: CommonTileParams = Depends(),
+    mosaic_params: CommonMosaicParams = Depends(),
+):
+    """Read MosaicJSON tile"""
+    # TODO: Maybe use ``read_mosaic`` defined above (depending on cache behavior which is still TBD)
+    pixsel = pixel_selection.method()
+    timings = []
+    headers: Dict[str, str] = {}
+
+    with utils.Timer() as t:
+        with MosaicBackend(mosaic_params.mosaic_path) as mosaic:
+            assets = mosaic.tile(x=x, y=y, z=z)
+            if not assets:
+                raise TileNotFoundError(f"No assets found for tile {z}/{x}/{y}")
+    timings.append(("Read-mosaic", t.elapsed))
+
+    tilesize = 256 * scale
+
+    # Rio-tiler-mosaic uses an external ThreadPoolExecutor to process multiple assets at once but we want to use the
+    # executor provided by the event loop.  Instead of calling ``rio_tiler_mosaic.mosaic.mosaic_tiler`` directly we will
+    # transcribe the code here and use the executor provided by the event loop.  This also means we define this function
+    # as a coroutine (even though nothing that is called is a coroutine), since the event loop's executor isn't
+    # available in normal ``def`` functions.
+    # https://github.com/cogeotiff/rio-tiler-mosaic/blob/master/rio_tiler_mosaic/mosaic.py#L37-L102
+    _tiler = partial(
+        cogeoTiler,
+        tile_x=x,
+        tile_y=y,
+        tile_z=z,
+        tilesize=tilesize,
+        indexes=image_params.indexes,
+        # expression=image_params.expression, # TODO: Figure out why expression kwarg doesn't work
+        nodata=image_params.nodata,
+        **image_params.kwargs,
+    )
+    futures = [run_in_threadpool(_tiler, asset) for asset in assets]
+
+    semaphore = asyncio.Semaphore(int(os.getenv("MOSAIC_CONCURRENCY", 10)))
+
+    with utils.Timer() as t:
+        async with semaphore:
+            for fut in asyncio.as_completed(futures):
+                try:
+                    tile, mask = await fut
+                except Exception:
+                    # Gracefully handle exceptions
+                    continue
+
+                tile = numpy.ma.array(tile)
+                tile.mask = mask == 0
+                pixsel.feed(tile)
+                if pixsel.is_done:
+                    break
```

Comment:
maybe it's time to re-write rio-tiler-mosaic using async?

Reply:
Maybe.  I'd want to see a comparison between threadpool executor and asyncio first.  Asyncio is certainly better for FastAPI but may differ for lambda etc.